packages/injectable/ogre-tools-injectable.d.ts

```diff
@@ -6,118 +6,128 @@ declare module '@ogre-tools/injectable' {
-          lifecycle: ILifecycle<InstantiationParam, string | number | symbol>
-        }
-    ),
-  ): Injectable<Instance, InjectionInstance, InstantiationParam>;
+    options: GetInjectableOptions<InjectionInstance, InjectionTokenInstance, InstantiationParam>,
+  ): Injectable<InjectionInstance, InjectionTokenInstance, InstantiationParam>;
 
   export function getInjectionToken<
-    Instance,
+    InjectionInstance,
     InstantiationParam = void,
   >({
     id: string,
-  }): InjectionToken<Instance, InstantiationParam>;
-
-  type AsyncReturnable<IsAsync extends boolean, Instance> = IsAsync extends true
-    ? Promise<Instance>
-    : Instance;
-
-  export interface Inject<IsAsync extends boolean> {
-    <Instance>(
-      key: Injectable<Instance, unknown, void> | InjectionToken<Instance, void>,
-    ): AsyncReturnable<IsAsync, Instance>;
-    <Instance, InstantiationParam>(
-      key: Injectable<Instance, unknown, InstantiationParam> | InjectionToken<Instance, InstantiationParam>,
+  }): InjectionToken<InjectionInstance, InstantiationParam>;
+
+  type AsyncReturnable<IsAsync extends boolean, InjectionInstance> = IsAsync extends true
+    ? Promise<InjectionInstance>
+    : InjectionInstance;
+
+  interface Inject<IsAsync extends boolean> {
+    <InjectionInstance>(
+      key: Injectable<InjectionInstance, unknown, void> | InjectionToken<InjectionInstance, void>,
+    ): AsyncReturnable<IsAsync, InjectionInstance>;
+    <InjectionInstance, InstantiationParam>(
+      key: Injectable<InjectionInstance, unknown, InstantiationParam> | InjectionToken<InjectionInstance, InstantiationParam>,
+      param: InstantiationParam,
+    ): AsyncReturnable<IsAsync, InjectionInstance>;
+  }
+
+  interface InjectMany<IsAsync extends boolean> {
+    <InjectionInstance>(
+      key: Injectable<InjectionInstance, unknown, void> | InjectionToken<InjectionInstance, void>,
+    ): AsyncReturnable<IsAsync, InjectionInstance[]>;
+    <InjectionInstance, InstantiationParam>(
+      key: Injectable<InjectionInstance, unknown, InstantiationParam> | InjectionToken<InjectionInstance, InstantiationParam>,
       param: InstantiationParam,
-    ): AsyncReturnable<IsAsync, Instance>;
+    ): AsyncReturnable<IsAsync, InjectionInstance[]>;
   }
 
   interface DiContainerForInjection<TReturnAsPromise extends boolean> {
     inject: Inject<TReturnAsPromise>;
-
-    injectMany<Instance, InstantiationParam = void>(
-      injectionToken: InjectionToken<Instance, InstantiationParam>,
-      instantiationParameter: InstantiationParam,
-    ): AsyncReturnable<TReturnAsPromise, Instance[]>;
+    injectMany: InjectMany<TReturnAsPromise>;
   }
 
-  export interface ILifecycle<InstantiationParam, Key extends string | number | symbol> {
-    getInstanceKey: (di: DiContainer, params: InstantiationParam) => Key;
+  export interface ILifecycle<InstantiationParam> {
+    getInstanceKey: (di: DiContainer, params: InstantiationParam) => string | number | symbol | object | bigint;
   }
```

Comment:
Note: Type of an instance key can be permitted to be anything, not just `string | number | symbol | object | bigint`.

Reply:
The only thing missing from this is `boolean`, `undefined`, and `null` which I think is probably best to not allow.